Add support for correlation-based convective BC to linear FV physics

GiudGiud opened the pull request:
Builds on @grmnptr 's PR
Will rebase soon

refs https://github.com/idaholab/moose/issues/30595 
https://github.com/idaholab/moose/issues/30281

GiudGiud committed:
Clean up error message

GiudGiud committed:
Add support for convective BCs in linear FV heat transfer physics
Fix up functor material logic
refs #30281

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2898057/), step Docs: sync website on 5a03386 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30592/site)

This comment will be updated on new commits.

GiudGiud committed:
Add a test for physics + heat flux BCs
refs #30281

GiudGiud committed:
Add support for heat flux BCs  in linear FV heat transfer physics
refs #30281

GiudGiud committed:
Safer treatment of mesh velocity subtraction in RC UOs and flow rate PPs
refs #30595

lindsayad commented on framework/src/timesteppers/IterationAdaptiveDT.C:
```diff
@@ -575,8 +576,18 @@ IterationAdaptiveDT::acceptStep()
+
+  // Use the total number of iterations for multi-system
+  for (const auto i : make_range(_fe_problem.numNonlinearSystems()))
+    _nl_its += _fe_problem.getNonlinearSystemBase(/*nl_sys=*/i).nNonlinearIterations();
```
Does this time stepper even work if `_nl_its` is 0 as would be the case with all linear systems? I would not include this commit in this PR if this functionality is not truly tested

GiudGiud replied:
It works to exponentially increase the time step size instead of seg-faulting, which is what I wanted :D

lindsayad replied:
Seems like an error would be even better

GiudGiud replied:
this is really commonly used so it should work. I fixed it beyond the seg fault now

GiudGiud replied:
ok still some work there

GiudGiud replied:
seems fine in the test. But not with PIMPLE ugh

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/cht/flow-around-square-linear-fluidonly-physics.i:
```diff
@@ -0,0 +1,155 @@
+  []
+[]
+
+[Functions]
+  [boundary_value]
+    type = ConstantFunction
+    value = 350
+  []
+[]
```
Could just add this into your `GenericFunctorMaterial`. Or also make `htc` a `ConstantFunction`

GiudGiud replied:
turns out this was not the right way to do this. I followed the same convection as in the nonlinear physics

GiudGiud replied:
so this is deleted

grmnptr commented on modules/navier_stokes/src/postprocessors/VolumetricFlowRate.C:
```diff
@@ -57,7 +58,9 @@ VolumetricFlowRate::VolumetricFlowRate(const InputParameters & parameters)
                ? &getUserObject<RhieChowFaceFluxProvider>("rhie_chow_user_object")
                : nullptr),
-    _subtract_mesh_velocity(getParam<bool>("subtract_mesh_velocity"))
+    _subtract_mesh_velocity(isParamValid("subtract_mesh_velocity")
+                                ? getParam<bool>("subtract_mesh_velocity")
+                                : _fe_problem.haveDisplaced())
```
I think this should actually be an error if we don't have mesh displacements and someone wants subtract mesh velocity.

grmnptr replied:
I mean we will subtract 0, but I think we should expect the user to have this right.

grmnptr reviewed (changes requested):
Are you planning to add a test case for the model that has both the fluid and the solid domains?

GiudGiud commented on modules/navier_stokes/src/physics/WCNSLinearFVFluidHeatTransferPhysics.C:
```diff
@@ -342,6 +342,41 @@ WCNSLinearFVFluidHeatTransferPhysics::addEnergyWallBC()
+      getProblem().addLinearFVBC(
+          bc_type, var_name + "_heatflux_" + wall_boundaries[bc_ind], params);
+    }
+    else if (_energy_wall_types[bc_ind] == "convection")
```
I think this is fine. I ll add the enum description instead

GiudGiud committed:
Address reviews:
- describe more what convection HT wall type entails

Also, use same syntax for NL and LinearFV convective HT

lindsayad commented on framework/src/timesteppers/IterationAdaptiveDT.C:
```diff
@@ -117,7 +121,8 @@ IterationAdaptiveDT::IterationAdaptiveDT(const InputParameters & parameters)
     _growth_factor(getParam<Real>("growth_factor")),
     _cutback_factor(getParam<Real>("cutback_factor")),
-    _nl_its(declareRestartableData<unsigned int>("nl_its", 0)),
+    _solver_its(declareRestartableData<unsigned int>("nl_its", 0)),
```
```suggestion
    _solver_its(declareRestartableData<unsigned int>("solver_its", 0)),
```

lindsayad commented on framework/include/timesteppers/IterationAdaptiveDT.h:
```diff
@@ -100,7 +100,9 @@ class IterationAdaptiveDT : public TimeStepper, public PostprocessorInterface
   const Real & _cutback_factor;
 
-  /// Number of nonlinear iterations in previous solve
+  /// Number of solver iterations in previous solve. Linear systems contribute both to _l_its and _solver_its
```
I think even more doxygen here would be good. Linear iteration within KSPSolves of an outer SNESSolve is still some kind of solver iteration

lindsayad replied:
I see in the docstring for `optimal_iterations` you use `outer iterations`. I like that name better, so my suggestion would be `solver_its` -> `outer_its`

lindsayad commented on framework/src/timesteppers/IterationAdaptiveDT.C:
```diff
@@ -512,24 +517,25 @@ IterationAdaptiveDT::computeAdaptiveDT(Real & dt, bool allowToGrow, bool allowTo
+  const std::string ite_type = (_fe_problem.numLinearSystems() == 0) ? "nl" : "solver";
 
-  if (allowToGrow && (_nl_its < growth_nl_its && _l_its < growth_l_its))
+  if (allowToGrow && (_solver_its < growth_nl_its && _l_its < growth_l_its))
```
whether we use `_solver_its` or `_outer_its`, let's rename `growth_nl_its` to match

lindsayad commented on test/tests/time_steppers/iteration_adaptive/tests:
```diff
@@ -62,6 +62,16 @@
+    csvdiff = 'adapt_linear_systems_out.csv'
+    requirement = "The system shall support the ability to adapt the time step size based on the number of iterations from linear systems."
+    # Number of linear iterations varies a lot in parallel
+    max_parallel = 1
```
This is a warning bell for me. Tests that are essentially iteration counters are a recipe for failure with changing architectures (e.g. GPU), slight tweaks in PETSc algorithms, etc. I would be ok with using something like `-pc_type lu -pc_factor_mat_solver_type mumps` which we can reasonably trust to always give the same number of iterations

GiudGiud replied:
we would need it to take more than 1 iteration though.

I might be able to cook something up with a Convergence object

GiudGiud replied:
ok removed. 
@joshuahansel  I ended up modifying the convergence code to be able to use it here. 
What I did is not the only option we could:
- spring up a linear convergence parameter, that we then use both for linear systems and KSP-nested instead nonlinear systems
- instead of calling nonlinearxyz -> solverxyz (what I did, following what Peter did for adding linear systems), we could call it: outerSolver, SolverSystem, ... . Is SolverConvergence the best name? Maybe not. SolverSystemConvergence is a tad more accurate

joshuahansel replied:
Discussed offline; current plan would be to keep `nonlinear_convergence` and add `linear_convergence`, as this would be more consistent with Executioner params: `nonlinear_convergence` is ignored if you're not doing any nonlinear solve. I think it's easier for users to understand than `outer/inner_convergence`.

GiudGiud replied:
outer/inner is not the best for sure. 
The only concern with opening up "linear" is that this means more than my use case and we ll just put an error outside of my use case for now

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2909297/), step Clang format on b4eb2cc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30592/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30592/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7cc8bd7530c04b73ef16019251092c5a42403fdc`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2910836/), step Clang format on 2f9f185 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30592/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30592/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7cc8bd7530c04b73ef16019251092c5a42403fdc`

GiudGiud commented:
Canceled tests are unrelated

loganharbour commented:
> Canceled tests are unrelated

Not appropriate to ignore 14 tests

GiudGiud commented:
sure I ll do that now.

grmnptr reviewed (commented):
How difficult would it be to add the example which has the solid solve as well?

grmnptr reviewed (approved):
Ahh okay, If we want to do an MSR with a reflector we will need it, and that is a main target application for this BC/coupling. Would be nice if we could use physics for that too but we can also roll without the physics for now.

GiudGiud commented:
you can couple the regular syntax with the Physics syntax.
The main application for my PR was the MSFL loop where we could have deployed this without solving for the solid region.

Thanks for the review

lindsayad commented on framework/src/utils/PetscSupport.C:
```diff
@@ -354,6 +354,55 @@ petscNonlinearConverged(SNES /*snes*/,
+
+      // TODO: find a KSP code that works better for this case
+    case Convergence::MooseConvergenceStatus::CONVERGED:
+      *reason = KSP_CONVERGED_RTOL_NORMAL;
```
I already did it for SNES in https://gitlab.com/petsc/petsc/-/merge_requests/8005. I can do it for KSP now

GiudGiud replied:
awesome. I think it would help make this cleaner
